### PR TITLE
Add profiler to full finetune recipes

### PR DESCRIPTION
**Pull request by gau-nernst**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses. #1279 

#### Changelog
What are the changes made in this PR?

Add profiler to `full_finetune_distributed.py` and `full_finetune_single_device.py`. Basically copied over from LoRA recipes.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](../tests/torchtune) for any new functionality
- [ ] update [docstrings](../docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

Just like the LoRA recipe, to profile, run:

```
tune run full_finetune_single_device \
  --config llama2/7B_full_low_memory \
  profiler.enabled=True \
  profiler.output_dir=./profile-test
```

Then we can open the trace under `profile-test/iteration_12` with something like https://ui.perfetto.dev

image

And it reveals that, in my setup, most of the time is spent transferring data back and forth between GPU and CPU for paged Adam.

### CODE REVIEW

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -254,6 +255,70 @@ def setup(self, cfg: DictConfig) -> None:
             self._steps_per_epoch = self.max_steps_per_epoch
         self.global_step = self.epochs_run * self._steps_per_epoch
 
+        # Set up profiler, returns DummyProfiler (nullcontext object with no-op `step` method)
+        # if cfg is missing profiler key or if `cfg.profiler.enabled = False`
+        self._profiler = self._setup_profiler(cfg.get(PROFILER_KEY, None))
+
+    def _setup_profiler(
+        self, cfg_profiler: Optional[DictConfig]
```
Comment (SalmanMohammadi) [inline recipes/full_finetune_distributed.py:263]: nit:
```suggestion
        self, cfg_profiler: Optional[DictConfig] = None
```

File: recipes/full_finetune_distributed.py [hunk 2]
```diff
@@ -254,6 +255,70 @@ def setup(self, cfg: DictConfig) -> None:
...
+    ) -> Union[torch.profiler.profile, DummyProfiler]:
+        """
+        Parses the `profiler` section of top-level `cfg` and sets up profiler
+
+        Args:
+            cfg_profiler (DictConfig): `profiler` section of the top-level `cfg` (the main config passed to `recipe.main`)
```
Comment (SalmanMohammadi) [inline recipes/full_finetune_distributed.py:269]: nit: 
```suggestion
            cfg_profiler (Optional[DictConfig]): `profiler` section of the top-level `cfg` (the main config passed to `recipe.main`)
```

Comment (SalmanMohammadi) [reply recipes/full_finetune_distributed.py:269]: I think this is the case in all the recipes. Sorry about our annoying linter : )

**Comment by SalmanMohammadi** [comment]
Minor nit to appease our linter but LGTM. Thanks for getting this out!!

**Comment by SalmanMohammadi** [comment]
cc @ebsmothers @joecummings to double check whether a quick example profile would be good to see

**Comment by gau-nernst** [comment]
Fix the linter issue. Sorry I didn't install pre-commit before making the first commit, so it slips through subsequent checks. Manually run pylintdoc to double check and no issue now. Also add the full finetune low memory trace to the PR description.

**Review by ebsmothers — APPROVED** [review]
Looks great, thank you for adding this!

**Pull request merged**